README.md
# My Usage

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-07-3pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-07-3pm us-east1 us-east1-b
   https://cc8fb569cd450d0-dot-us-east1.pipelines.googleusercontent.com
./destroy.sh mlops-poc-337511 mlops-july-07-3pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-07-1pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-07-1pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-07-1pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-11pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-11pm us-east1 us-east1-b
   https://4469656e93fe3549-dot-us-east1.pipelines.googleusercontent.com
./destroy.sh mlops-poc-337511 mlops-july-05-11pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-6pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-6pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-05-6pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-3pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-3pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-05-3pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-2pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-2pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-05-2pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-1pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-1pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-05-1pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-8am us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-8am us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-05-8am us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-1am us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05-1am us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-05-1am us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-05 us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-05 us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-05 us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-11pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-11pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-04-11pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-7pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-7pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-04-7pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-630pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-630pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-04-630pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-430pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-04-430pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-04-430pm us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-with-mlflow us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-with-mlflow us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-with-mlflow us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret gcp-with-mlflow us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret gcp-with-mlflow us-east1 us-east1-b
./destroy.sh mlops-poc-337511 gcp-with-mlflow us-east1 us-east1-b

# Setting up an MLOps Environment with Cloud Composer and MLflow on Google Cloud

We describe how to provision an MLOps environment using [Cloud Composer](https://cloud.google.com/composer)
and [MLflow](https://mlflow.org/) on Google Cloud. This environment enables deploying and executing
ML continuous training pipelines, as well as managing and tracking ML experiments, metadata, and artifacts.

## Environment overview

The diagram shows the overall MLOps environment

![architecture](../../images/mlops-composer-mlflow.png)

The proposed MLOps environment consists of the following components:

| Component         | Description |
|-----------------|-------------|
|[AI Platform Notebooks](https://cloud.google.com/ai-platform-notebooks) | A managed JupyterLab instances for ML experimentation. The instance uses a custom container with MLflow installed and Cloud Proxy to connect to Cloud SQL
|[Cloud Composer](https://cloud.google.com/composer) | A managed [Airflow](https://airflow.apache.org/) service to deploy, schedule, and orchestration  run ML training pipelines. It runs on [Google Kubernetes Engine](https://cloud.google.com/kubernetes-engine).
|[MLflow](https://mlflow.org/) | An open source platform for managing the ML ifecycle.  In this environment, MLflow used for tracking ML metadata and artifacts.
|[Cloud SQL](https://cloud.google.com/sql/docs) | A managed relational database service. In this environment, Cloud SQL used as a backend for the [MLflow tracking](https://www.mlflow.org/docs/latest/tracking.html#tracking).
|[Cloud Storage](https://cloud.google.com/storage) | A simple, reliable, and highly-durable object store. In this environment, Cloud Storage is used as an artifact store, where the outputs of ML pipeline steps are saved.
|[Cloud Build](https://cloud.google.com/cloud-build)| A managed service for executing CI/CD routines on Google Cloud infrastructure. In this environment, Cloud Build is used to build the AI Notebooks instance custom container image, as well as  ML model training runtimes.
|[Container Registry](https://cloud.google.com/container-registry)| A single place for a team to store and manage Docker images. In this environment, Container Registry is used to store the container images that are produced by Cloud Build.

In this environment, the MLflow Tracking server is a used in locally in three services.

1. AI Platform Notebook instance(s) - accessed by local Notebooks.
2. Cloud Composer GKE Cluster - accessed by Airflow through K8s cluster internal IP. MLflow is deployed as a pod in the GKE cluster.
3. AI Platform Training - accessed locally by the ML training code.

All the MLflow tracking services share the same backend MySQL database hosted in Cloud SQL, thanks to MLflow Tracking Service stateless operation.
This setup reduces the interconnection complexity.

In addition, [Cloud SQL proxy](https://cloud.google.com/sql/docs/mysql/sql-proxy)
is used to allow MLflow to connect to the Cloud SQL instance.


## Before you begin

In the following steps we assume that you already have a [GCP Project](https://cloud.google.com/cloud-resource-manager), with
[billing enabled](https://cloud.google.com/billing/docs/how-to/modify-project). You will also need
[Project Editor](https://cloud.google.com/iam/docs/understanding-roles) permission to provision this environment.

Update to the latest gcloud command

   ```
   sudo gcloud components update
   ```

## Installation Script

Provisioning of the environment has been automated with the [install.sh](install.sh) script.
The installation script performs the following steps:

1. Enables necessary APIs
2. Creates a Cloud Storage bucket as an artifact store
3. Creates Cloud SQL instance
4. Provisions a Composer cluster environment in GKE
5. Deploys MLflow server to Composer GKE cluster

### 1. Enable the required APIs

In addition to the [services enabled by default](https://cloud.google.com/service-usage/docs/enabled-service), 
the following additional services must be enabled:

- Compute Engine
- Container Registry
- Cloud Build
- Cloud Composer
- Cloud Dataflow
- Cloud SQL Admin
- Notebooks

### 2. Create a Cloud Storage bucket

A Cloud Storage bucket is required as an ML **artifact store**. The artifacts produced by the various ML steps, such as 
data splits, trained models, evaluation metrics, will be saved in this bucket.

The Cloud Storage bucket name will be set to **<DEPLOYMENT_NAME>-artifacts**.

### 3. Provisioning a Cloud SQL instance

A Cloud SQL instance is used as an MLflow backend to store ML metadata and pointers to artifacts stored
in the Cloud Storage bucket.

The Cloud SQL instance name will be set to **<DEPLOYMENT_NAME>-sql**, and the default **root** user password will be set to <SQL_PASSWORD>.
A database named **mlflow** will be created in the Cloud SQL, MySQL instance.

For more information, see [Cloud SQL setup](https://cloud.google.com/sql/docs/mysql>) and about user and security management,
see [Creating and managing MySQL users](https://cloud.google.com/sql/docs/mysql/create-manage-users).


### 4. Provisioning Cloud Composer

Cloud Composer will run ML training pipelines, implemented in Airflow, on a managed environment.
Cloud Composer installation created a GKE cluster in the `REGION` and `ZONE` provided.

The Cloud Composer cluster name will be set to **<DEPLOYMENT_NAME>-af**

By default, the `--machine-type` is set to `n1-standard-2`, `--node-count` is set to 3, and `--python-version` is set to 3.

After the Cloud Composer is provisioned, Python packages in the [composer-requirements.txt](composer-requirements.txt) file are installed
to the Composer runtime. The file includes MLflow, SciPy, and Scikit-learn. For more information,
see [Creating environments](https://cloud.google.com/composer/docs/how-to/managing/creating) in the Cloud Composer documentation.

> Note that the ML steps like data pre-processing and model training and evaluation, where large compute
> resources are required advised to
> to execute in their appropriate GCP managed services (e.g. Dataflow and AI Platform),
> rather than in the Composer runtime. However, some small steps might be executed with the Composer runtime,
> like comparing evaluation metrics and logging information to MLflow tracking. Thus, some libraries need
> to be installed in the Composer runtime.
---


### 5. Deploying MLflow server to Composer GKE cluster

Deploying and running MLflow as a pod to the Composer GKE cluster requires the following:

1. MLflow server will need to connect to the Cloud SQL to store the ML metadata.
The connection will be possible via a [Cloud SQL Proxy](https://cloud.google.com/sql/docs/mysql/sql-proxy).
Therefore, script creates a service account `sql-proxy-access@PROJECT_ID.iam.gserviceaccount.com`,
and downloads the account key to `sql-access.json` file.
Key will be used for the [Cloud SQL Proxy authorization](https://cloud.google.com/sql/docs/mysql/authorize-proxy).
The service account is granted a single `cloudsql.client` IAM role, which enables the connection between proxy and
MySQL server, but it does not restrict database level permissions. Database access is managed by general MySQL user name based
authorization.

2. A Docker container image with MLflow is created and pushed to Container Registry.
The container image uses this [Dockerfile](mlflow-helm/docker/Dockerfile), where `mlflow server`
command is the entry point.
Along with the main MLflow image, a side-car container will be created to be a proxy server to expose
[MLflow Tracking web UI](https://www.mlflow.org/docs/latest/tracking.html#tracking-ui).

3. [Helm](https://helm.sh/) is used to deploy the MLflow container to the Composer's GKE cluster.
Helm compiles Kubernetes application configuration and deploys all components.
The helm templates for the MLflow server installation are found in [mlflow-helm](mlflow-helm) folder.

### 6. Build the common ML container image

Services and Jupyter notebook in ML container have access to provisioned infrastructure components such as 
Cloud SQL and MLflow Tracking service.
Connection URIs and other settings are propagated via environment variables.
Environment setting for the AI Platform Notebooks instance is created and stored 
in file in Cloud Storage, to be used in provisioning an AI Platform Notebooks instance:

   ```
    cat > custom-ml-image/notebook-env.txt << EOF
    MLFLOW_GCS_ROOT_URI=${GCS_BUCKET_NAME}
    MLFLOW_SQL_CONNECTION_STR=mysql+pymysql://${SQL_USERNAME}:${SQL_PASSWORD}@127.0.0.1:3306/mlflow
    MLFLOW_SQL_CONNECTION_NAME=$(gcloud sql instances describe ${CLOUD_SQL} --format="value(connectionName)")
    MLFLOW_EXPERIMENTS_URI=${GCS_BUCKET_NAME}/experiments
    MLFLOW_TRACKING_URI=http://127.0.0.1:80
    MLFLOW_TRACKING_EXTERNAL_URI="https://"$(kubectl describe configmap inverse-proxy-config -n mlflow | grep "googleusercontent.com")
    MLOPS_COMPOSER_NAME=${DEPLOYMENT_NAME}-af
    MLOPS_REGION=${REGION}
    EOF

    gsutil cp custom-ml-image/notebook-env.txt ${gs://$DEPLOYMENT_NAME-artifacts}
    rm custom-ml-image/notebook-env.txt
   ```

MLflow local service instance connects Cloud SQL service which is defined in 'init.sh' leveraging these variables.



changes.txt
https://1ae0e2ebaf78e03c-dot-us-east1.pipelines.googleusercontent.com

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-12-1248pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-12-1248pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-12-1248pm us-east1 us-east1-b

ERROR: (gcloud) Cannot specify --python-version with Composer 2.X or greater.

Error on line: 163
Caused by: gcloud composer environments create $COMPOSER_NAME --location=$REGION --zone=$ZONE --airflow-configs=core-dags_are_paused_at_creation=True --disk-size=50GB --image-version=composer-2.0.19-airflow-2.2.5 --machine-type=n1-standard-2 --node-count=3 --python-version=3 --enable-ip-alias
That returned exit status: 1
Aborting...
Install Python packages to Cloud Composer...
ERROR: (gcloud.composer.environments.update) NOT_FOUND: No such environment found: projects/mlops-poc-337511/locations/us-east1/environments/mlops-july-12-1156am-af
Error on line: 163
Caused by: gcloud composer environments update $COMPOSER_NAME --update-pypi-packages-from-file=composer-requirements.txt --location=$REGION
That returned exit status: 1
Aborting...
Cloud Composer is available: mlops-july-12-1156am-af

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-12-1156am us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-12-1156am us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-12-1156am us-east1 us-east1-b


ERROR: Cannot install SQLAlchemy==1.4.0, apache-airflow==2.2.5 and mlflow==1.26.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested SQLAlchemy==1.4.0
    mlflow 1.26.1 depends on sqlalchemy
    apache-airflow 2.2.5 depends on sqlalchemy<1.4.0 and >=1.3.18

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
The command '/bin/sh -c apt-get update && apt-get install -y --no-install-recommends mysql-client kubectl &&     rm -rf /var/lib/apt/lists/* &&     pip3 install pip install SQLAlchemy==1.4.0 pandas==1.0.4 click==7.0 PyMySQL==0.9.3 google-cloud-storage==2.4.0 mlflow==1.26.1 apache-airflow==2.2.5 PyYAML==5.3.1 cloudml-hypertune &&     curl -o /usr/local/bin/cloud_sql_proxy https://dl.google.com/cloudsql/cloud_sql_proxy.linux.amd64 &&     chmod +x /usr/local/bin/cloud_sql_proxy &&     chmod +x /init.sh &&     chmod +x /entrypoint.sh' returned a non-zero code: 1
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: step exited with non-zero status: 1

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-1124pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-1124pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-11-1124pm us-east1 us-east1-b


root@b8033b7b32aa:/home/jupyter# 2022/07/11 20:44:59 ERROR mlflow.cli: Error initializing backend store
2022/07/11 20:44:59 ERROR mlflow.cli: No module named 'sqlalchemy.future'
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/mlflow/cli.py", line 426, in server
    initialize_backend_stores(backend_store_uri, default_artifact_root)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/server/handlers.py", line 259, in initialize_backend_stores
    _get_tracking_store(backend_store_uri, default_artifact_root)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/server/handlers.py", line 244, in _get_tracking_store
    _tracking_store = _tracking_store_registry.get_store(store_uri, artifact_root)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/tracking/_tracking_service/registry.py", line 39, in get_store
    return self._get_store_with_resolved_uri(resolved_store_uri, artifact_uri)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/tracking/_tracking_service/registry.py", line 49, in _get_store_with_resolved_uri
    return builder(store_uri=resolved_store_uri, artifact_uri=artifact_uri)
  File "/opt/conda/lib/python3.7/site-packages/mlflow/server/handlers.py", line 110, in _get_sqlalchemy_store
    from mlflow.store.tracking.sqlalchemy_store import SqlAlchemyStore
  File "/opt/conda/lib/python3.7/site-packages/mlflow/store/tracking/sqlalchemy_store.py", line 11, in <module>
    from sqlalchemy.future import select
ModuleNotFoundError: No module named 'sqlalchemy.future'

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-1037am us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-1037am us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-11-1037am us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-837am us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-837am us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-11-837am us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-1232am us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-11-1232am us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-11-1232am us-east1 us-east1-b

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-10-1114pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-10-1114pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-10-1114pm us-east1 us-east1-b

# SQLAlchemy==1.3.24
./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-10-1051pm us-east1 us-east1-b
./install.sh mlops-poc-337511 mlflow_db_secret mlops-july-10-1051pm us-east1 us-east1-b
./destroy.sh mlops-poc-337511 mlops-july-10-1051pm us-east1 us-east1-b


Does not have minimum availability	- CrashLoopBackOff

./set-env-vars.sh mlops-poc-337511 mlflow_db_secret mlops-july-10-1111am us-east1 us-east1-b

install.sh
#!/bin/bash

# Copyright 2020 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#            http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Check command line parameters
if [[ $# < 2 ]]; then
  echo 'USAGE:  ./install.sh PROJECT_ID SQL_PASSWORD [DEPLOYMENT_NAME=mlops] [REGION=us-central1] [ZONE=us-central1-a]'
  echo 'PROJECT_ID      - GCP project Id'
  echo 'SQL_PASSWORD    - Password to connect database'
  echo 'DEPLOYMENT_NAME - Short name prefix of infrastructure element and folder names, like SQL instance, Cloud Composer name'
  echo 'REGION          - A GCP region across the globe. Best to select one of the nearest where Cloud AI Platform available.'
  echo 'ZONE            - A zone is an isolated location within a region. Available Regions and Zones: https://cloud.google.com/compute/docs/regions-zones'
  exit 1
fi

# Set project and default constants
export PROJECT_ID=${1}
export SQL_PASSWORD=${2}
export DEPLOYMENT_NAME=${3:-mlops}
export REGION=${4:-us-central1}
export ZONE=${5:-us-central1-a}
# export USE_GKE_GCLOUD_AUTH_PLUGIN=True
# gcloud components install gke-gcloud-auth-plugin

export SQL_USERNAME="root"
# Set calculated infrastucture and folder names
export GCS_BUCKET_NAME="gs://$DEPLOYMENT_NAME-artifacts"
export ML_IMAGE_URI="gcr.io/$PROJECT_ID/$DEPLOYMENT_NAME-mlimage:latest"

# Set up a global error handler
err_handler() {
    echo "Error on line: $1"
    echo "Caused by: $2"
    echo "That returned exit status: $3"
    echo "Aborting..."
    return $3
}

if [[ -z "${PROJECT_ID}" ]]; then
  echo PROJECT_ID is not set
  exit 1
fi

if [[ -z "${SQL_USERNAME}" ]]; then
  echo SQL_USERNAME is not set
  exit 1
fi

if [[ -z "${SQL_PASSWORD}" ]]; then
  echo SQL_PASSWORD is not set
  exit 1
fi

if [[ -z "${DEPLOYMENT_NAME}" ]]; then
  echo DEPLOYMENT_NAME is not set
  exit 1
fi

if [[ -z "${REGION}" ]]; then
  echo REGION is not set
  exit 1
fi

if [[ -z "${ZONE}" ]]; then
  echo ZONE is not set
  exit 1
fi

if [[ -z "${GCS_BUCKET_NAME}" ]]; then
  echo GCS_BUCKET_NAME is not set
  exit 1
fi

if [[ -z "${ML_IMAGE_URI}" ]]; then
  echo ML_IMAGE_URI is not set
  exit 1
fi

trap 'err_handler "$LINENO" "$BASH_COMMAND" "$?"' ERR

tput setaf 3; echo Creating environment
echo Setup started at:
date
tput setaf 7

# Set calculated infrastucture names
CLOUD_SQL="$DEPLOYMENT_NAME-sql"
COMPOSER_NAME="$DEPLOYMENT_NAME-af"
MLFLOW_IMAGE_URI="gcr.io/${PROJECT_ID}/$DEPLOYMENT_NAME-mlflow"
MLFLOW_PROXY_URI="gcr.io/${PROJECT_ID}/inverted-proxy"

# 1. Enable services

echo "Enabling all required services..."

gcloud services enable \
  cloudbuild.googleapis.com \
  sourcerepo.googleapis.com \
  container.googleapis.com \
  compute.googleapis.com \
  composer.googleapis.com \
  containerregistry.googleapis.com \
  dataflow.googleapis.com \
  sqladmin.googleapis.com \
  notebooks.googleapis.com

echo "Required services enabled."
echo

#2. Create a Cloud Storage bucket

echo "Creating GCS bucket for artifacts..."
if ! gsutil list "$GCS_BUCKET_NAME"; then
gsutil mb -p $PROJECT_ID -l $REGION $GCS_BUCKET_NAME
fi
echo "GCS bucket available: $GCS_BUCKET_NAME"
echo

# 3. Provisioning a Cloud SQL instance

if [[ $(gcloud sql instances list --filter="$CLOUD_SQL" --format='value(name)') != "$CLOUD_SQL" ]]; then
    echo "Provisioning Cloud SQL..."
    gcloud sql instances create $CLOUD_SQL --tier=db-g1-small --region=$REGION
    gcloud sql databases create mlflow --instance=$CLOUD_SQL
    gcloud sql users set-password $SQL_USERNAME --host=% --instance=$CLOUD_SQL --password=$SQL_PASSWORD
fi
MLFLOW_SQL_CONNECTION_NAME=$(gcloud sql instances describe $CLOUD_SQL --format="value(connectionName)")
echo "Cloud SQL is available: $MLFLOW_SQL_CONNECTION_NAME"

# 4. Provisioning Cloud Composer

# Creating Cloud Composer
if [[ $(gcloud composer environments list --locations=$REGION --filter="$COMPOSER_NAME" --format='value(name)') != "$COMPOSER_NAME" ]]; then
    echo "Provisioing Cloud Composer..."
    gcloud composer environments create $COMPOSER_NAME \
    --location=$REGION \
    --zone=$ZONE \
    --airflow-configs=core-dags_are_paused_at_creation=True \
    --disk-size=50GB \
    --image-version=composer-1.19.2-airflow-2.1.4 \
    --machine-type=n1-standard-2 \
    --node-count=3 \
    --python-version=3 \
    --enable-ip-alias

# Installing Python packages to Composer
  echo "Install Python packages to Cloud Composer..."
  gcloud composer environments update $COMPOSER_NAME \
    --update-pypi-packages-from-file=composer-requirements.txt \
    --location=$REGION
fi
echo "Cloud Composer is available: $COMPOSER_NAME"
echo


# 5. Deploying MLflow server to Composer GKE cluster

echo "Provisioning MLflow Tracking server..."

# Set local Kubernetes configuration to connect to Composer GKE cluster
echo "Setting configuration to connect to Composer GKE cluster..."
# 'goog-composer-environment' label is set to $COMPOSER_NAME
GKE_CLUSTER=$(gcloud container clusters list --limit=1 --zone=$ZONE --filter="resourceLabels.goog-composer-environment=$COMPOSER_NAME" --format="value(name)")
gcloud container clusters get-credentials $GKE_CLUSTER --zone $ZONE --project $PROJECT_ID
echo "GKE Cluster is: $GKE_CLUSTER"
echo

# Create service account
SA_EMAIL=sql-proxy-access@$PROJECT_ID.iam.gserviceaccount.com
if [[ $(gcloud iam service-accounts list --filter="$SA_EMAIL" --format='value(email)') != "$SA_EMAIL" ]]; then
    echo "Create new service account: $SA_EMAIL"
    gcloud iam service-accounts create sql-proxy-access --format='value(email)' --display-name="Cloud SQL access for sql proxy"
fi

# Download service account key
if [[ -e mlflow-helm/sql-access.json ]]; then
    echo "Service account key already exists: mlflow-helm/sql-access.json"
else
    gcloud iam service-accounts keys create mlflow-helm/sql-access.json --iam-account=$SA_EMAIL
    cp mlflow-helm/sql-access.json custom-ml-image/
fi

# Set role to the service account
echo "Set cloudsql.client role to the service account..."
gcloud projects add-iam-policy-binding $PROJECT_ID \
--member serviceAccount:$SA_EMAIL \
--role roles/cloudsql.client
echo "IAM policy binding is added."

# Build MLflow docker image
echo "Build MLflow Docker container image..."
gcloud builds submit mlflow-helm/docker --timeout 15m --tag ${MLFLOW_IMAGE_URI}:latest
echo "MLflow Docker container image is built: ${MLFLOW_IMAGE_URI}:latest"

# Build MLflow UI proxy image
echo "Build MLflow UI proxy container image..."
gcloud builds submit mlflow-helm/proxy --timeout 15m --tag ${MLFLOW_PROXY_URI}:latest
echo "MLflow UI proxy container image is built: ${MLFLOW_PROXY_URI}:latest"

# Initializing Helm's tiller
echo "Initializing Helm environment..."
# kubectl create serviceaccount --namespace kube-system tiller
kubectl --namespace kube-system create serviceaccount tiller
kubectl create clusterrolebinding tiller-cluster-rule --clusterrole=cluster-admin --serviceaccount=kube-system:tiller

# Using fix K8s namespace: 'mlflow' for MLflow
echo "Create mlfow namespace to the GKE cluster..."
kubectl create namespace mlflow || echo "'mlflow' namespace already exists"

# Deploying mlflow using helm
echo "Deploying mlflow helm configuration..."
helm install mlflow --namespace mlflow \
--set images.mlflow=$MLFLOW_IMAGE_URI \
--set images.proxyagent=$MLFLOW_PROXY_URI \
--set defaultArtifactRoot=$GCS_BUCKET_NAME/experiments \
--set backendStore.mysql.host="127.0.0.1" \
--set backendStore.mysql.port="3306" \
--set backendStore.mysql.database="mlflow" \
--set backendStore.mysql.username=$SQL_USERNAME \
--set backendStore.mysql.password=$SQL_PASSWORD \
--set cloudSqlConnection.name=$MLFLOW_SQL_CONNECTION_NAME \
mlflow-helm

# Generate command for debug:
echo helm template mlflow --namespace mlflow --set images.mlflow=$MLFLOW_IMAGE_URI --set images.proxyagent=$MLFLOW_PROXY_URI --set defaultArtifactRoot=$GCS_BUCKET_NAME/experiments --set backendStore.mysql.host="127.0.0.1" --set backendStore.mysql.port="3306" --set backendStore.mysql.database="mlflow" --set backendStore.mysql.user=$SQL_USERNAME --set backendStore.mysql.password=$SQL_PASSWORD --set cloudSqlConnection.name=$MLFLOW_SQL_CONNECTION_NAME --output-dir './yamls' mlflow-helm

MLFLOW_SQL_CONNECTION_STR="mysql+pymysql://$SQL_USERNAME:$SQL_PASSWORD@127.0.0.1:3306/mlflow"

echo "Waiting for MLflow Tracking server provisioned"

# External URL to Mlflow
MLFLOW_TRACKING_EXTERNAL_URI="https://"
# Internal access from Composer to Mlflow
MLFLOW_URI_FOR_COMPOSER=="http://"
echo "wait 5 seconds..."
sleep 5s
while [ "$MLFLOW_TRACKING_EXTERNAL_URI" == "https://" ] || [ "$MLFLOW_URI_FOR_COMPOSER" == "http://" ]
do
  MLFLOW_TRACKING_EXTERNAL_URI="https://"$(kubectl describe configmap inverse-proxy-config -n mlflow | grep "googleusercontent.com")
  MLFLOW_URI_FOR_COMPOSER="http://"$(kubectl get svc -n mlflow mlflow -o jsonpath='{.spec.clusterIP}{":"}{.spec.ports[0].port}')
done

echo "MLflow External URL:"
echo $MLFLOW_TRACKING_EXTERNAL_URI

echo "MLflow Composer URL:"
echo $MLFLOW_URI_FOR_COMPOSER

echo "MLflow Tracking server provisioned"
echo
echo 6. Build the common ML container image

# init.sh will be executed durring trainer image containerization
cat > custom-ml-image/init.sh << EOF
#!/bin/bash
export MLFLOW_GCS_ROOT_URI=${GCS_BUCKET_NAME}
export MLFLOW_SQL_CONNECTION_STR=${MLFLOW_SQL_CONNECTION_STR}
export MLFLOW_SQL_CONNECTION_NAME=${MLFLOW_SQL_CONNECTION_NAME}
export MLFLOW_EXPERIMENTS_URI=${GCS_BUCKET_NAME}/experiments
export MLFLOW_TRACKING_URI=http://127.0.0.1:80
export MLFLOW_TRACKING_EXTERNAL_URI=${MLFLOW_TRACKING_EXTERNAL_URI}
export MLOPS_COMPOSER_NAME=${COMPOSER_NAME}
export MLOPS_REGION=${REGION}
export ML_IMAGE_URI=${ML_IMAGE_URI}

/usr/local/bin/cloud_sql_proxy -dir=/var/run/cloud-sql-proxy -instances=${MLFLOW_SQL_CONNECTION_NAME}=tcp:3306 -credential_file=/usr/local/bin/sql-access.json &
sleep 45s
mlflow server --host=127.0.0.1 --port=80 --backend-store-uri=${MLFLOW_SQL_CONNECTION_STR} --default-artifact-root=${GCS_BUCKET_NAME}/experiments &
EOF

# Custom ML image URI needed for projecting a new trainer job from Notebooks or Airflow 
gcloud builds submit custom-ml-image --timeout 15m --tag ${ML_IMAGE_URI}

# Note: MLflow provisioning takes minutes. After the mlimage creation it should be available.

# Add MLflow URI to Cloud Composer as environment variable
echo "Add MLflow URI to Cloud Composer as environment variable..."
gcloud composer environments update $COMPOSER_NAME \
  --update-env-variables MLFLOW_TRACKING_URI=$MLFLOW_URI_FOR_COMPOSER,MLFLOW_GCS_ROOT_URI=$GCS_BUCKET_NAME \
  --location $REGION \
  --async
echo

# Create connection info which will be used as environment variables inside the Notebook instance.
cat > custom-ml-image/notebook-env.txt << EOF
MLFLOW_GCS_ROOT_URI=${GCS_BUCKET_NAME}
MLFLOW_SQL_CONNECTION_STR=${MLFLOW_SQL_CONNECTION_STR}
MLFLOW_SQL_CONNECTION_NAME=${MLFLOW_SQL_CONNECTION_NAME}
MLFLOW_EXPERIMENTS_URI=${GCS_BUCKET_NAME}/experiments
MLFLOW_TRACKING_URI=http://127.0.0.1:80
MLFLOW_TRACKING_EXTERNAL_URI=${MLFLOW_TRACKING_EXTERNAL_URI}
MLOPS_COMPOSER_NAME=${COMPOSER_NAME}
MLOPS_REGION=${REGION}
ML_IMAGE_URI=${ML_IMAGE_URI}
EOF

gsutil cp custom-ml-image/notebook-env.txt $GCS_BUCKET_NAME
rm custom-ml-image/notebook-env.txt
tput setaf 3;
echo "MLflow UI can be accessed externally at the below URI:"
echo $MLFLOW_TRACKING_EXTERNAL_URI
tput setaf 7;

echo "Enviornment is provisioned successfully."
